pkg/spacebinding/spacebindinglister.go

```diff
@@ -0,0 +1,65 @@
+package spacebinding
+
+import (
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	errs "github.com/pkg/errors"
+)
+
+// Lister allows to list all spacebindings for a given space.
+type Lister struct {
+	ListSpaceBindingsFunc func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error)
+	GetSpaceFunc          func(spaceName string) (*toolchainv1alpha1.Space, error)
+}
+
+func NewLister(listSpaceBindingsFunc func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error), getSpaceFunc func(spaceName string) (*toolchainv1alpha1.Space, error)) *Lister {
+	return &Lister{
+		ListSpaceBindingsFunc: listSpaceBindingsFunc,
+		GetSpaceFunc:          getSpaceFunc,
+	}
+}
+
+// ListForSpace it recursively searches up. It returns all the SBs from this space and from all the parent spaces of that space.
+// It doesn't search SBs in the child spaces.
+func (l *Lister) ListForSpace(space *toolchainv1alpha1.Space, foundBindings []toolchainv1alpha1.SpaceBinding) ([]toolchainv1alpha1.SpaceBinding, error) {
+	parentBindings, err := l.ListSpaceBindingsFunc(space.Name)
+	if err != nil {
+		return foundBindings, err
+	}
+
+	// spaceBindings is the list that will be returned, it will contain either parent and child merged or just the "parent" ones retrieved above.
+	foundBindings = mergeSpaceBindings(foundBindings, parentBindings)
+
+	// no parent space,
+	// let's return list of bindings accumulated since here ...
+	if space.Spec.ParentSpace == "" {
+		return foundBindings, nil
+	}
+
+	// fetch parent space and recursively keep going ...
+	parentSpace, err := l.GetSpaceFunc(space.Spec.ParentSpace)
+	if err != nil {
+		// Error reading the object
+		return foundBindings, errs.Wrap(err, "unable to get parent-space")
+	}
+
+	return l.ListForSpace(parentSpace, foundBindings)
+}
+
+func mergeSpaceBindings(foundBindings []toolchainv1alpha1.SpaceBinding, parentBindings []toolchainv1alpha1.SpaceBinding) []toolchainv1alpha1.SpaceBinding {
+	// if both list are not empty, we have to merge them.
+	// roles for the same username on SpaceBindings will override those on parentSpaceBindings,
+	// so let's remove them from parentSpaceBinding before merging the two lists.
+	for _, spaceBinding := range foundBindings {
+		// iterate from back to front, so you don't have to worry about indexes that are deleted.
+		for i := len(parentBindings) - 1; i >= 0; i-- {
+			if parentBindings[i].Spec.MasterUserRecord == spaceBinding.Spec.MasterUserRecord {
+				parentBindings = append(parentBindings[:i], parentBindings[i+1:]...)
+				break
+			}
+		}
+	}
+	// merge lists now that there are no duplicates
+	// and just use the TypeMeta and ListMeta objects from the spaceBinding list
+	// since only the Items field is relevant from this object
+	return append(foundBindings, parentBindings...)
```

Comment:
looking at the PR in registration-service - we could modify this function to make sure that we return always only one SpaceBinding per username, so we don't have to care about the situation when one user would have accidentally more SpaceBindings, right?

Reply:
I can create a map with the username as key, but if there is more then one spacebinding per username I guess we want to be spot that out and throw an error. We can return an error from here and let the client (registration-serivce and host-operator) decide what todo with it in theory. Or as we do right now, the client decides if it's ok or not to have more then one spacebinding per username.

Reply:
the question is where to show the error. IMO, this and the registration-service are the wrong places for throwing an error. If two SpaceBindings are created for one MUR, then the generic Space controller logic will apply templates of  both roles to the given username, which means that the user will really have both roles applied in the namespaces.
Let's do the check in SpaceBindingRequest API - if there is already a SpaceBinding for the given MUR, then the SBR should wail with appropriate condition

Reply:
btw, feel free to create a new story for it and do it as a follow-up of the whole effort. We don't have to do it now.

Reply:
> Theoretically, the MUR uniqueness should be ensured by using always the same naming format (which is something we do) so the risk is very minimal.

Right, I think this case is already covered in the SBR controller implicitly by using the same names for the `spacebinding` resource. So in theory if the user tries to create a SBR for the same MUR and Space, the SBR controller will throw an error here: https://github.com/codeready-toolchain/host-operator/blob/8e85ee74a26db8c7c8f7aebaac53449e7b688e13/controllers/spacebindingrequest/spacebindingrequest_controller.go#L286-L296

Is this what you were thinking about or is any other think to cover that I'm missing?

Reply:
> btw, feel free to create a new story for it and do it as a follow-up of the whole effort. 

I have created ASC-430 for this small improvement.